### PR TITLE
add support for distributed data parallel training

### PR DESCRIPTION
**Pull request by ImahnShekhzadeh**
This PR adds support for distributed data parallel (DDP) and replaces `DataParallel` with `DistributedDataParallel` in `train_cifar.py`, which can be used via the flag `parallel`. To achieve this, the code is refactored, and the flags `master_addr` and `master_port` are added.

I tested the changes, on a single GPU, I get an FID of 3.74 (with the OT-CFM method), on two GPUs with DDP, I get an FID of 3.81.

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **test your PR locally** with `pytest` command?
- [x] Did you **run pre-commit hooks** with `pre-commit run -a` command?

### CODE REVIEW

**Comment by kilianFatras** [comment]
Hi, thank you for your contribution! 

I had an internal implementation with fabric form litghtning but I like to rely only on PyTorch for this example. I need some time to review it (a few days/weeks). I will come back to it soon.

File: examples/images/cifar10/compute_fid.py [hunk 1]
```diff
@@ -51,7 +51,7 @@
 # Load the model
 PATH = f"{FLAGS.input_dir}/{FLAGS.model}/{FLAGS.model}_cifar10_weights_step_{FLAGS.step}.pt"
 print("path: ", PATH)
-checkpoint = torch.load(PATH)
+checkpoint = torch.load(PATH, map_location=device)
```
Comment (kilianFatras) [inline examples/images/cifar10/compute_fid.py:54]: would the code work if you run inference over 1 GPU while the network was trained on 2?

File: examples/images/cifar10/train_cifar10.py [hunk 2]
```diff
@@ -169,5 +175,17 @@ def train(argv):
                 )
 
 
+def main(argv):
+    # get world size (number of GPUs)
+    world_size = int(os.getenv("WORLD_SIZE", 1))
```
Comment (kilianFatras) [inline examples/images/cifar10/train_cifar10.py:180]: I feel like world_size is not really a good name. What do you think of calling it 'total_number_of_gpus' or something in that spirit?

File: examples/images/cifar10/train_cifar10.py [hunk 3]
```diff
@@ -65,6 +67,12 @@ def train(argv):
         FLAGS.save_step,
     )
 
+    if FLAGS.parallel and world_size > 1:
+        # When using `DistributedDataParallel`, we need to divide the batch
+        # size ourselves based on the total number of GPUs of the current node.
+        FLAGS.batch_size = int(FLAGS.batch_size / world_size)
```
Comment (kilianFatras) [inline examples/images/cifar10/train_cifar10.py:73]: I don't like overwriting the FLAGS's batch size. Can you rename it 'batch_size_per_gpu' or just 'batch_size? and update line 91?

File: examples/images/cifar10/train_cifar10.py [hunk 4]
```diff
@@ -81,7 +89,8 @@ def train(argv):
     dataloader = torch.utils.data.DataLoader(
         dataset,
         batch_size=FLAGS.batch_size,
-        shuffle=True,
+        sampler=DistributedSampler(dataset) if FLAGS.parallel else None,
+        shuffle=False if FLAGS.parallel else True,
```
Comment (kilianFatras) [inline examples/images/cifar10/train_cifar10.py:93]: hum. I am rather unsure about this. where do you shuffle the data then?

**Review by kilianFatras — CHANGES_REQUESTED** [review]
Thank you for this nice contribution! My main concern is about the data shuffling. I think we should keep shuffling them. Maybe I am missing something and happy to learn about it.

File: examples/images/cifar10/compute_fid.py (hunk 1 above)
Comment (ImahnShekhzadeh) [reply examples/images/cifar10/compute_fid.py:54]: Yep. I'd trained on 2 GPUs, but done inference on only 1 GPU. If I remember correctly, without `map_location=device`, an error was thrown when doing inference on only 1 GPU, while training was done on multiple GPUs.

File: examples/images/cifar10/train_cifar10.py (hunk 2 above)
Comment (ImahnShekhzadeh) [reply examples/images/cifar10/train_cifar10.py]: I think it's quite common to call this variable `world_size`, but I admit that `total_num_gpus` or sth alike is more descriptive, so I changed the variable name accordingly

File: examples/images/cifar10/train_cifar10.py (hunk 4 above)
Comment (ImahnShekhzadeh) [reply examples/images/cifar10/train_cifar10.py:95]: Fair point. I found this warning in the [PyTorch docs](https://pytorch.org/docs/stable/data.html): 
image

So in my current implementation, `train_sampler.set_epoch(epoch)` is missing, which I will add now.

File: examples/images/cifar10/compute_fid.py (hunk 1 above)
Comment (kilianFatras) [reply examples/images/cifar10/compute_fid.py:54]: Awesome. Thank you for trying this.

File: examples/images/cifar10/train_cifar10.py (hunk 4 above)
Comment (kilianFatras) [reply examples/images/cifar10/train_cifar10.py:95]: Perfect. Once you have finished your change, I will run the code myself. Once I get it working, I will merge the PR.

Final question, can you try to load and run the existing checkpoints? I just want to be sure that people can reproduce our results. Thx.

Comment (ImahnShekhzadeh) [reply examples/images/cifar10/train_cifar10.py:98]: Ok, I refactored the training loop to use `num_epochs` instead of `FLAGS.total_steps`, since `sampler.set_epoch(epoch)` uses an epoch count. However, I think we need to change more than this. The PyTorch warning I pasted above mentions that we need to use `sampler.set_epoch(epoch)` "**before** creating the `DataLoader` iterator", but right now, the data loader iterator is created **once** before the training loop:
```python
from utils_cifar import infiniteloop

datalooper = infiniteloop(dataloader)
```
The way I would change this is by having a training loop like this:
```python
with trange(num_epochs, dynamic_ncols=True) as epoch_pbar:
    for epoch in epoch_pbar:
        epoch_pbar.set_description(f"Epoch {epoch + 1}/{num_epochs}")
        if sampler is not None:
            sampler.set_epoch(epoch)
        for batch_idx, data in enumerate(dataloader): 
                # step += 1  # tricky
                optim.zero_grad()
                x1 = data.to(rank)  # old: `x1 = next(datalooper)`
                [...]
```
Is this fine by you? IMO, what is a bit tricky is to handle the `step` counter correctly (based on which checkpoints are saved and some samples during training are generated). In a distributed setup, we'll have several processes running in parallel, and thus, we would probably save checkpoints and images multiple times (once per process)? However, since the filenames do not reflect the process ID, one process would also overwrite the files of the other. What do you think?

Comment (ImahnShekhzadeh) [reply examples/images/cifar10/train_cifar10.py:98]: About your question: When you say "existing checkpoints", which ones do you mean? I had once run the training and generation of samples on one GPU and gotten an FID of `3.8` (which is only slightly worse than the `3.5` you report).

**Comment by kilianFatras** [comment]
I like the new changes. @atong01 do you mind having a look? I also think it would be great to keep the original train_cifar10.py. 

While I like this code, it is slightly more complicated than the previous one. So I would keep both. The idea of this package is that any master student can easily understand it in 1hour. @ImahnShekhzadeh can you rename this file train_cifar10_ddp.py please? and re-add the previous file? Thanks

**Comment by ImahnShekhzadeh** [comment]
>  @ImahnShekhzadeh can you rename this file train_cifar10_ddp.py please? and re-add the previous file? Thanks

Done

**Comment by atong01** [comment]
LGTM. Thanks for the contribution @ImahnShekhzadeh
